onnxruntime/contrib_ops/cpu/bert/ngram_repeat_block.h

```diff
@@ -0,0 +1,84 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#pragma once
+
+#include "core/common/common.h"
+#include "core/framework/op_kernel.h"
+#include "core/platform/threadpool.h"
+
+namespace onnxruntime {
+namespace contrib {
+class NGramRepeatBlock : public OpKernel {
+ public:
+  explicit NGramRepeatBlock(const OpKernelInfo& info) : OpKernel(info) {
+    ORT_ENFORCE(info.GetAttr<int64_t>("ngram_size", &ngram_size_).IsOK());
+  }
+
+  Status Compute(OpKernelContext* context) const override {
+    const Tensor* input_ids = context->Input<Tensor>(0);
+    const Tensor* scores = context->Input<Tensor>(1);
+    Tensor* output = context->Output(0, scores->Shape());
+
+    const auto* scores_source = static_cast<const float*>(scores->DataRaw());
+    auto* scores_target = static_cast<float*>(output->MutableDataRaw());
+    if (scores_source != scores_target) {
+      memcpy(scores_target, scores_source, scores->Shape().Size() * sizeof(float));
+    }
+
+    const auto& input_ids_dims = input_ids->Shape().GetDims();
+    const auto& scores_dims = scores->Shape().GetDims();
+    ORT_ENFORCE(input_ids_dims.size() == 2);
+    ORT_ENFORCE(scores_dims.size() == 2);
+    int64_t batch_size = input_ids_dims[0];
+    int64_t cur_len = input_ids_dims[1];
+    ORT_ENFORCE(scores_dims[0] == batch_size);
+    int64_t vocab_size = scores_dims[1];
+
+    if (cur_len + 1 < ngram_size_ || ngram_size_ <= 0) {
+      return Status::OK();
```

Comment:
Not clear if need to throw something when ngram_size_ <= 0?

Reply:
nice catch, maybe the check should go in initialization to disallow non-positive ngram_size.